tracing-opentelemetry/src/layer.rs

```diff
@@ -163,37 +202,131 @@ where
+    /// use tracing_subscriber::layer::SubscriberExt;
+    /// use tracing_subscriber::Registry;
+    ///
+    /// // The probability sampler can be used to export a percentage of spans
+    /// let sampler = sdk::Sampler::Probability(0.33);
+    ///
+    /// // Create a layer with the configured sampler
+    /// let otel_layer = tracing_opentelemetry::layer().with_sampler(sampler);
     ///
     /// // Use the tracing subscriber `Registry`, or any other subscriber
     /// // that impls `LookupSpan`
     /// let subscriber = Registry::default()
     ///     .with(otel_layer);
     /// # drop(subscriber);
     /// ```
-    pub fn with_tracer(tracer: T) -> Self {
+    pub fn with_sampler<Sampler>(self, sampler: Sampler) -> Self
+    where
+        Sampler: api::Sampler + 'static,
+    {
+        OpenTelemetryLayer {
+            sampler: Box::new(sampler),
+            ..self
+        }
+    }
+
+    /// Construct a new layer with the specified [`Tracer`] and [`Sampler`].
+    ///
+    /// [`Tracer`]: https://docs.rs/opentelemetry/latest/opentelemetry/api/trace/tracer/trait.Tracer.html
+    /// [`Sampler`]: https://docs.rs/opentelemetry/latest/opentelemetry/api/trace/sampler/trait.Sampler.html
+    fn new<Sampler>(tracer: T, sampler: Sampler) -> Self
+    where
+        Sampler: api::Sampler + 'static,
+    {
         OpenTelemetryLayer {
             tracer,
+            sampler: Box::new(sampler),
+            id_generator: sdk::IdGenerator::default(),
             get_context: WithContext(Self::get_context),
             _registry: marker::PhantomData,
         }
     }
 
+    /// Retrieve the parent OpenTelemetry [`SpanContext`] from the current
+    /// tracing [`span`] through the [`Registry`]. This [`SpanContext`]
+    /// links spans to their parent for proper hierarchical visualization.
+    ///
+    /// [`SpanContext`]: https://docs.rs/opentelemetry/latest/opentelemetry/api/trace/span_context/struct.SpanContext.html
+    /// [`span`]: https://docs.rs/tracing/latest/tracing/struct.Span.html
+    /// [`Registry`]: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/struct.Registry.html
+    fn parent_span_context(
+        &self,
+        attrs: &Attributes<'_>,
+        ctx: &Context<'_, S>,
+    ) -> Option<api::SpanContext> {
+        // If a span is specified, it _should_ exist in the underlying `Registry`.
+        if let Some(parent) = attrs.parent() {
+            let span = ctx.span(parent).expect("Span not found, this is a bug");
+            let mut extensions = span.extensions_mut();
+            extensions
+                .get_mut::<api::SpanBuilder>()
+                .map(|builder| build_span_context(builder, self.sampler.as_ref()))
+        // Else if the span is inferred from context, look up any available current span.
+        } else if attrs.is_contextual() {
+            ctx.current_span().id().and_then(|span_id| {
+                let span = ctx.span(span_id).expect("Span not found, this is a bug");
```

Comment:
Note that this could probably be simplified to use [`Context::lookup_current`](https://docs.rs/tracing-subscriber/0.2.5/tracing_subscriber/layer/struct.Context.html#method.lookup_current) which is a utility function for getting the current span's ID and looking up its stored data.

Reply:
ah wasn't aware of `lookup_current`, will switch.